Fail to build exe.wixproj

[10日前ぐらいから](https://github.com/kai2nenobu/win-python-installer/actions/runs/3389749361)Python 3.7～3.10のMSIビルドに失敗するようになった。

直接のエラーはWIXで`exe.wixproj`をビルドする際に発生している。

```
Project "D:\a\win-python-installer\win-python-installer\cpython\Tools\msi\bundle\releaselocal.wixproj" (1) is building "D:\a\win-python-installer\win-python-installer\cpython\Tools\msi\exe\exe.wixproj" (8) on node 1 (Rebuild target(s)).
CoreClean:
  Creating directory "D:\a\win-python-installer\win-python-installer\cpython\PCbuild\obj\\38amd64_Release\msi_exe\".
_GenerateGuids:
  "D:\a\win-python-installer\win-python-installer\cpython\Tools\msi\..\..\PCbuild\\amd64\python.exe" -c "import uuid; print('\n'.join('{}={}'.format(i, uuid.uuid5(uuid.UUID('c8d9733e-a70c-43ff-ab0c-e26456f11083'), 'http://www.python.org/amd64/' + j)) for i,j in [('CoreUpgradeCode', '3.8/upgradecode'),('UpgradeCode', '3.8/upgradecode/exe'),('InstallDirectoryGuidSeed', '3.8/installdirectoryseed'),('PythonExeComponentGuid', '3.8/python.exe'),('PythonwExeComponentGuid', '3.8/pythonw.exe'),('RemoveLib2to3PickleComponentGuid', '3.8/lib2to3/pickles'),('CommonPythonRegComponentGuid', '3.8/registry'),('PythonRegComponentGuid', '3.8/registry/exe')]))" > "D:\a\win-python-installer\win-python-installer\cpython\PCbuild\obj\\38amd64_Release\msi_exe\exeguids.txt"
_MergeMiscNewsWithPython:
  "py" -m pip install -U blurb
  Unable to create process using 'C:\hostedtoolcache\windows\Python\3.11.0\x64\py"  -m pip install -U blurb': The system cannot find the file specified.


D:\a\win-python-installer\win-python-installer\cpython\Tools\msi\exe\exe.wixproj(45,9): error MSB3073: The command ""py" -m pip install -U blurb" exited with code 101.
Done Building Project "D:\a\win-python-installer\win-python-installer\cpython\Tools\msi\exe\exe.wixproj" (Rebuild target(s)) -- FAILED.
```

手元のPCでもPython 3.7とPython 3.11の両バージョンをインストールした上で、Python 3.8.15をビルドすると再現した。該当するエラーは以下の通り。

```
プロジェクト "C:\Users\kai2nenobu\ghq\github.com\python\cpython\Tools\msi\bundle\releaselocal.wixproj" (1) は、ノード 1 上に "C:\Users\kai2nenobu\ghq\github.com\python\cpython\Tools\msi\exe\exe.wixpro
j" (8) をビルドしています (Rebuild ターゲット)。
CoreClean:
  ...
_GenerateGuids:
  ...
_MergeMiscNewsWithPython:
  "py" -m pip install -U blurb
  Unable to create process using 'C:\Python311\py"  -m pip install -U blurb': ??????????????????

C:\Users\kai2nenobu\ghq\github.com\python\cpython\Tools\msi\exe\exe.wixproj(45,9): error MSB3073: コマンド ""py" -m pip install -U blurb" はコード 101 で終了しました。
プロジェクト "C:\Users\kai2nenobu\ghq\github.com\python\cpython\Tools\msi\exe\exe.wixproj" (Rebuild ターゲット) のビルドが終了しました -- 失敗。
```

`exe.wixproj`の該当するところを見ると、以下のようになっている。

```xml
    <Target Name="_MergeMiscNewsWithBlurb" AfterTargets="PrepareForBuild" Condition="$(Blurb) != '' and !Exists('$(PySourcePath)Misc\NEWS')">
        <Exec Command="$(Blurb) merge -f &quot;$(BuildPath)NEWS.txt&quot;" WorkingDirectory="$(PCbuild)" />
    </Target>
    
    <Target Name="_MergeMiscNewsWithPython" AfterTargets="PrepareForBuild" Condition="$(Blurb) == '' and !Exists('$(PySourcePath)Misc\NEWS')">
        <ItemGroup>
            <HostPython Include="$(ExternalsDir)python*\tools\python.exe" />
            <HostPython Include="@(HostPython)" Condition="Exists(%(FullPath))" />
            <HostPython Include="py" Condition="@(HostPython) == ''" />
        </ItemGroup>
        <PropertyGroup>
            <HostPython>@(HostPython)</HostPython>
            <HostPython Condition="$(HostPython.Contains(';'))">$(HostPython.Remove($(HostPython.IndexOf(';'))))</HostPython>
        </PropertyGroup>
        <Exec Command="&quot;$(HostPython)&quot; -m pip install -U blurb" WorkingDirectory="$(PCbuild)" />
        <Exec Command="&quot;$(HostPython)&quot; -m blurb merge -f &quot;$(BuildPath)NEWS.txt&quot;" WorkingDirectory="$(PCbuild)" />
    </Target>
```

`"&quot;$(HostPython)&quot; -m pip install -U blurb"`のところが実行されていて、`HostPython`の実体が`py`なので`"py" -m pip install -U blurb`というコマンドが実行されることになる。手元のコマンドプロンプトでコマンドを実行してみると、たしかに同じエラーになる。

```console
>"py" -m pip install -U blurb
Unable to create process using 'C:\Python311\py"  -m pip install -U blurb': ??????????????????
```

そもそも`"py"`単独で実行してもエラーになるので、Python Lancher自身の問題の可能性が高い。

```console
>"py"
Unable to create process using 'C:\Python311\py"': ??????????????????
```

<https://github.com/actions/runner-images/commit/8bdf5c36087ae1104905787ce0413d2eb355c1ab>を見ると、10日ぐらい前にGitHub-Hosted runnerの`windows-2019`にPython 3.11が追加されているので、時期もちょうど符合する。

--- a/ci/build_installer.ps1
+++ b/ci/build_installer.ps1
@@ -48,7 +48,14 @@ if ($Ref -match '^v?3\.[78]') {
   Pop-Location
 }
 
-# Build installer
+# Avoid a build error for exe.wixproj between version 3.7 and 3.10 (ref. https://github.com/kai2nenobu/win-python-installer/issues/28)
+if ($Ref -match '^v?3\.([789]|10)') {
+  'Avoid a build error for exe.wixproj between version 3.7 and 3.10'
+  Push-Location cpython
+  git apply ..\patch\avoid_pylauncher_311_error.patch
+  Pop-Location
+}
 
+# Build installer
 cmd /c cpython\Tools\msi\buildrelease.bat @BuildOptions
 if ($LASTEXITCODE -gt 0) { exit $LASTEXITCODE }

--- a/ci/build_installer.ps1
+++ b/ci/build_installer.ps1
@@ -1,4 +1,4 @@
-
+﻿
 [CmdletBinding()]
 Param(
   [Parameter(Position=0)]
@@ -40,9 +40,9 @@ if ($Ref -notmatch '^v?3\.[67]') {
   python -m pip install -r ./cpython/Doc/requirements.txt
 }
 
-# Avoid a build error for version 3.7 and 3.8 (ref. https://github.com/kai2nenobu/win-python-installer/issues/6)
-if ($Ref -match '^v?3\.[78]') {
-  'Avoid Windows 11 SDK in branch 3.7 and 3.8'
+# Avoid a build error for version 3.7 (ref. https://github.com/kai2nenobu/win-python-installer/issues/6)
+if ($Ref -match '^v?3\.7') {
+  'Avoid Windows 11 SDK in branch 3.7'
   Push-Location cpython
   git -c user.name=dummy -c 'user.email=dummy@example.com' am ..\patch\avoid_win11_sdk.patch
   Pop-Location

--- a/ci/build_installer.ps1
+++ b/ci/build_installer.ps1
@@ -1,4 +1,4 @@
-﻿
+
 [CmdletBinding()]
 Param(
   [Parameter(Position=0)]
@@ -53,6 +53,8 @@ if ($Ref -match '^v?3\.([789]|10)') {
   'Avoid a build error for exe.wixproj between version 3.7 and 3.10'
   Push-Location cpython
   git apply ..\patch\avoid_pylauncher_311_error.patch
+  git add --update .
+  git -c user.name=dummy -c 'user.email=dummy@example.com' commit -m 'Avoid a build error for exe.wixproj between version 3.7 and 3.10'
   Pop-Location
 }
 

--- a/patch/avoid_pylauncher_311_error.patch
+++ b/patch/avoid_pylauncher_311_error.patch
@@ -1,0 +1,11 @@
+--- a/Tools/msi/exe/exe.wixproj
++++ b/Tools/msi/exe/exe.wixproj
+@@ -36,7 +36,7 @@
+         <ItemGroup>
+             <HostPython Include="$(ExternalsDir)python*\tools\python.exe" />
+             <HostPython Include="@(HostPython)" Condition="Exists(%(FullPath))" />
+-            <HostPython Include="py" Condition="@(HostPython) == ''" />
++            <HostPython Include="py.exe" Condition="@(HostPython) == ''" />
+         </ItemGroup>
+         <PropertyGroup>
+             <HostPython>@(HostPython)</HostPython>

